### PR TITLE
Feature 34 Sale - product/user information validation

### PR DESCRIPTION
**Pull request by berki0**
adding validation methods for Sale

### CODE REVIEW

File: src/main/java/jewellery/inventory/service/SaleService.java [hunk 1]
```diff
@@ -33,23 +36,53 @@ public SaleResponseDto createSale(SaleRequestDto saleRequestDto) {
...
+    throwExceptionIfProductIsSold(sale.getProducts());
+    throwExceptionIfUserNotSeller(sale.getProducts(), saleRequestDto.getSellerId());
+    throwExceptionIfProductIsPartOfAnotherProduct(sale.getProducts());
+
     Sale createdSale = saleRepository.save(sale);
     updateProductOwnersAndSale(sale.getProducts(), saleRequestDto.getBuyerId(), createdSale);
     return saleMapper.mapEntityToResponseDto(createdSale);
   }
 
-  private void throwExceptionIfProductOwnerNotSeller(List<Product> products, UUID sellerId) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/SaleService.java:42]: Why did we rename this? `ProductOwnerNotSeller` looks more concise to me

File: src/main/java/jewellery/inventory/service/SaleService.java [hunk 2]
```diff
@@ -33,23 +36,53 @@ public SaleResponseDto createSale(SaleRequestDto saleRequestDto) {
...
+    }
+  }
+
+  private void updateOwnerAndSubProducts(Product product, User newOwner, Sale sale) {
+    productService.updateProductOwnerAndSale(product, newOwner, sale);
+    if (product.getProductsContent() != null) {
+      List<Product> subProducts = product.getProductsContent();
+      for (Product subProduct : subProducts) {
+        updateOwnerAndSubProducts(subProduct, newOwner, sale);
+      }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/SaleService.java]: We don't need this? We moved the logic in ProductService.

File: src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java [hunk 3]
```diff
@@ -28,6 +28,30 @@
...
+
+  @NotNull
+  private static ResourceInUserRequestDto getResourceInUserRequestDto(
+      User user, PreciousStone preciousStone) {
+    ResourceInUserRequestDto resourceInUserRequestDto = new ResourceInUserRequestDto();
+    resourceInUserRequestDto.setUserId(user.getId());
+    resourceInUserRequestDto.setResourceId(preciousStone.getId());
+    resourceInUserRequestDto.setQuantity(20);
+    return resourceInUserRequestDto;
+  }
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java]: This should be moved down to the other object creation methods.

File: src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java [hunk 4]
```diff
@@ -32,6 +32,48 @@
...
+
+  @NotNull
+  private static ResourceInUserRequestDto getResourceInUserRequestDto(
+      User user, PreciousStone preciousStone) {
+    ResourceInUserRequestDto resourceInUserRequestDto = new ResourceInUserRequestDto();
+    resourceInUserRequestDto.setUserId(user.getId());
+    resourceInUserRequestDto.setResourceId(preciousStone.getId());
+    resourceInUserRequestDto.setQuantity(20);
+    return resourceInUserRequestDto;
+  }
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java]: Same with these

File: src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java [hunk 5]
```diff
@@ -137,24 +172,6 @@ void createSaleSuccessfully() {
...
-    ProductPriceDiscountRequestDto productPriceDiscountRequestDto =
-        new ProductPriceDiscountRequestDto();
-    productPriceDiscountRequestDto.setProductId(productResponse.getBody().getId());
-    productPriceDiscountRequestDto.setSalePrice(10000.0);
-    productPriceDiscountRequestDto.setDiscount(10.0);
-    List<ProductPriceDiscountRequestDto> list = new ArrayList<>();
-    list.add(productPriceDiscountRequestDto);
-    saleRequestDto.setProducts(list);
-    return saleRequestDto;
-  }
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java]: Something strange has happened with this method. Why is it moved in the beginning of the file?

**Review by VladoKat — CHANGES_REQUESTED** [review]

File: src/main/java/jewellery/inventory/service/SaleService.java (hunk 1 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/SaleService.java:42]: I did not make/request this change so no idea. Will fix. Should not it be SellerNotProductOwner?

File: src/main/java/jewellery/inventory/service/SaleService.java (hunk 2 above)
Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/SaleService.java]: Removed

File: src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java (hunk 4 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java]: The formatter did this... probably my formatter actually. Having some issues. Fixed

File: src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java (hunk 3 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java]: The formatter did this... probably my formatter actually. Having some issues. Fixed

File: src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java (hunk 5 above)
Comment (Alekseeybg) [reply src/test/java/jewellery/inventory/integration/SaleIntegrationTest.java]: The formatter did this... probably my formatter actually. Fixed it.

File: src/main/java/jewellery/inventory/service/SaleService.java (hunk 1 above)
Comment (VladoKat) [reply src/main/java/jewellery/inventory/service/SaleService.java:42]: yep, `SellerNotProductOwner` is better considering the name of the exception `UserNotOwner`

Comment (Alekseeybg) [reply src/main/java/jewellery/inventory/service/SaleService.java:42]: Fixed but github does not acknowledge the change...

**Review by VladoKat — APPROVED** [review]

**Pull request merged**